### PR TITLE
Load all data from index using scroll api

### PR DESCRIPTION
**Pull request by mukeshPP**

### CODE REVIEW

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java [hunk 1]
```diff
@@ -320,6 +325,46 @@ public void reAlias(@NotEmpty  String newIndex, @NotEmpty String aliasName) {
...
+                .setScroll(scroll);
+        SearchResponse searchResponse = searchRequestBuilder
+                .execute()
+                .actionGet();
+
+        String scrollId = searchResponse.getScrollId();
+        List<T> batchedResult = ElasticUtils.getResponse(searchResponse, klass);
+        List<T> totalResult = new ArrayList<>(batchedResult);
+        int count = 1;
+        while (!batchedResult.isEmpty()) {
```
Comment (koushikr) [inline src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:369]: Couple of things. Please add the description on how you are fetching the scroll results in batches on top of the method. Second, you can reduce this to a function-pointer instead into a different private method and define your recursion better than a primitive way such as this!

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java [hunk 2]
```diff
@@ -320,6 +325,46 @@ public void reAlias(@NotEmpty  String newIndex, @NotEmpty String aliasName) {
...
 
+    /**
+     * Load data of entire index based on batchSize using ES scroll API
+     * @param <T> Response class Type
+     * @param index Index to be loaded
+     * @param query jelastic query object
+     * @param batchSize Will tell ElasticClient the size of each fetch, should be <= 10000 for better performance
+     * @return List<T> list of all objects in that index
+     */
+    public <T> List<T> loadAll(String index, Query query, Integer batchSize, Class<T> klass) {
```
Comment (koushikr) [inline src/main/java/io/github/jelastic/core/repository/ElasticRepository.java]: Why is it Integer but not int? Why is it boxed?

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java [hunk 3]
```diff
@@ -320,6 +325,46 @@ public void reAlias(@NotEmpty  String newIndex, @NotEmpty String aliasName) {
...
+                .prepareSearch(index)
+                .setQuery(queryBuilder)
+                .setSize(batchSize)
+                .setScroll(scroll);
+        SearchResponse searchResponse = searchRequestBuilder
+                .execute()
+                .actionGet();
+
+        String scrollId = searchResponse.getScrollId();
+        List<T> batchedResult = ElasticUtils.getResponse(searchResponse, klass);
```
Comment (koushikr) [inline src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: How many results can this searchResponse contain? ElasticUtils.getResponse does casting, this is going to be really expensive, are you going to impose any size checks on this scroll method?

**Review by koushikr — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java (hunk 1 above)
Comment (mukeshPP) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:369]: 1. Description will be added.
2. Regarding using lambda for iteration, thing is I am not iterating **on** any iterable or stream. What I am doing is imitating infinite loop, unless all the data is read into the List. Once no data is returned by ES, the infinite loop terminates. Other way would be using `IntStream.iterate` to imitate infinite loop, I feel that would be overkill for such simple use case just for sake of readability .

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java (hunk 2 above)
Comment (mukeshPP) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java]: 1. Will be changed to `int`

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java (hunk 3 above)
Comment (mukeshPP) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: 1. It fetches in batches of 10k till all the data is read from ES
2. No size check in place as of now. It is unbounded. In our use case we are caching entire ES data into In-Memory cache. Which might hit 100k record at max.

Comment (koushikr) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: You might want to still impose a size check, this lib is NOT for your use-case, alone! It is a general purpose lib, and I would like for size check to be imposed from config, with a default to 10K.

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java (hunk 1 above)
Comment (koushikr) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:369]: I agree. Intstream.iterate may be an overkill.

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java (hunk 3 above)
Comment (mukeshPP) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: Added a configuration for maxResultSize, defaulting to 10k.

Comment (koushikr) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: Please don't do an assert, please use the JElasticException. And use that as a guard rail, please use the constant from the configuration in the comparator not a hardcoded value!

File: pom.xml [hunk 4]
```diff
@@ -6,7 +6,7 @@
     <groupId>io.github.jelastic</groupId>
     <artifactId>jelastic</artifactId>
     <packaging>jar</packaging>
-    <version>7.2.0-3-SNAPSHOT</version>
+    <version>7.2.0-4-SNAPSHOT</version>
```
Comment (koushikr) [inline pom.xml]: Also make it a release version, not snapshot. And also edit README.md

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java (hunk 3 above)
Comment (mukeshPP) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: Added exception if limit is reached.

File: pom.xml (hunk 4 above)
Comment (mukeshPP) [reply pom.xml]: Changed to release version.

File: src/main/java/io/github/jelastic/core/repository/ElasticRepository.java (hunk 3 above)
Comment (koushikr) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: Why would you do that inside the while loop? A guard rail is supposed to work at the beginning of a method. Pl. make sure you do it before you do anything else. Pl. understand why you are throwing this exception - this is to mean, look I breached my size so I'm not performing this action - which is the narrative! I don't see your code doing it. Pl. think before the MR!

Comment (mukeshPP) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: How would I know if the fetched result is exceeding a specified limit unless I read the data from ES ?

Comment (koushikr) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: Thanks for making the change. The idea of guard rail is to protect you from things going wrong, not as a reactive check.

Comment (koushikr) [reply src/main/java/io/github/jelastic/core/repository/ElasticRepository.java:366]: Thanks for making the change. The idea of a guard rail is to protect you from doing any wrong, not to check reactively. This check now works.

**Review by koushikr — APPROVED** [review]

**Pull request merged**